### PR TITLE
Split expression names from SmtEngine

### PR DESCRIPTION
**Pull request by ajreynol**
Towards splitting SmtEngine / deleting SmtEnginePrivate.

### CODE REVIEW

File: src/smt/expr_names.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+namespace CVC4 {
+namespace smt {
+
+/**
+ * This utility is responsible for maintaining names for expressions.
+ */
+class ExprNames
+{
+  typedef std::unordered_map<Node, Node, NodeHashFunction> NodeToNodeHashMap;
```
Comment (nafur) [inline src/smt/expr_names.h:33]:
```suggestion
```
It seems this typedef is never actually used.
It we want to keep it:
```suggestion
  using NodeToNodeHashMap = std::unordered_map<Node, Node, NodeHashFunction>;
```
See [the guidelines](https://github.com/CVC4/CVC4/wiki/Code-Style-Guidelines#aliases) (and I very much support this)

File: src/smt/expr_names.h [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+ public:
+  ExprNames(context::UserContext* u);
+  ~ExprNames();
+  /**
+   * Get expression name.
+   *
+   * Return true if given expressoion has a name in the current context.
```
Comment (nafur) [inline src/smt/expr_names.h:40]:
```suggestion
   * Return true if given expression has a name in the current context.
```

File: src/smt/expr_names.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+   */
+  bool getExpressionName(Node e, std::string& name) const;
+
+  /**
+   * Set name of given expression 'e' to 'name'.
+   *
+   * This information is user-context-dependent.
+   * If 'e' already has a name, it is overwritten.
+   */
+  void setExpressionName(Node e, const std::string& name);
```
Comment (nafur) [inline src/smt/expr_names.h:51]:
```suggestion
  void setExpressionName(const Node& e, const std::string& name);
```

File: src/smt/expr_names.h [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+ public:
+  ExprNames(context::UserContext* u);
+  ~ExprNames();
+  /**
+   * Get expression name.
+   *
+   * Return true if given expressoion has a name in the current context.
+   * If it returns true, the name of expression 'e' is stored in 'name'.
+   */
+  bool getExpressionName(Node e, std::string& name) const;
```
Comment (nafur) [inline src/smt/expr_names.h:43]:
```suggestion
  bool getExpressionName(const Node& e, std::string& name) const;
```

File: src/smt/expr_names.cpp [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+namespace CVC4 {
+namespace smt {
+
+ExprNames::ExprNames(context::UserContext* u)
+    : d_exprNames(u)
+{
+}
+ExprNames::~ExprNames() {}
+
+void ExprNames::setExpressionName(Node e, const std::string& name) {
```
Comment (nafur) [inline src/smt/expr_names.cpp:26]:
```suggestion
void ExprNames::setExpressionName(const Node& e, const std::string& name) {
```

File: src/smt/expr_names.cpp [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+  d_exprNames[e] = name;
+}
+
+bool ExprNames::getExpressionName(Node e, std::string& name) const {
```
Comment (nafur) [inline src/smt/expr_names.cpp:30]:
```suggestion
bool ExprNames::getExpressionName(const Node& e, std::string& name) const {
```

File: src/smt/expr_names.cpp [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+  context::CDHashMap< Node, std::string, NodeHashFunction >::const_iterator it = d_exprNames.find(e);
```
Comment (nafur) [inline src/smt/expr_names.cpp:31]: Knowing full well that you might not like it :-)
```suggestion
  auto it = d_exprNames.find(e);
```

File: src/smt/expr_names.h [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+/**
+ * This utility is responsible for maintaining names for expressions.
+ */
+class ExprNames
+{
+  typedef std::unordered_map<Node, Node, NodeHashFunction> NodeToNodeHashMap;
+ public:
+  ExprNames(context::UserContext* u);
+  ~ExprNames();
```
Comment (nafur) [inline src/smt/expr_names.h:36]:
```suggestion
```
-> [rule of zero](https://en.cppreference.com/w/cpp/language/rule_of_three)

File: src/smt/expr_names.cpp [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+#include "smt/expr_names.h"
+
+namespace CVC4 {
+namespace smt {
+
+ExprNames::ExprNames(context::UserContext* u)
+    : d_exprNames(u)
+{
+}
+ExprNames::~ExprNames() {}
```
Comment (nafur) [inline src/smt/expr_names.cpp:24]:
```suggestion
```
-> [rule of zero](https://en.cppreference.com/w/cpp/language/rule_of_three)

File: src/smt/smt_engine.cpp [hunk 10]
```diff
@@ -3219,13 +3199,13 @@ CVC4::SExpr SmtEngine::getOption(const std::string& key) const
   return SExpr::parseAtom(d_options.getOption(key));
 }
 
-bool SmtEngine::getExpressionName(Expr e, std::string& name) const {
-  return d_private->getExpressionName(e, name);
+bool SmtEngine::getExpressionName(Node e, std::string& name) const {
```
Comment (nafur) [inline src/smt/smt_engine.cpp:3202]:
```suggestion
bool SmtEngine::getExpressionName(const Node& e, std::string& name) const {
```

File: src/smt/smt_engine.cpp [hunk 11]
```diff
@@ -3219,13 +3199,13 @@ CVC4::SExpr SmtEngine::getOption(const std::string& key) const
...
+  return d_exprNames->getExpressionName(e, name);
 }
 
-void SmtEngine::setExpressionName(Expr e, const std::string& name) {
+void SmtEngine::setExpressionName(Node e, const std::string& name) {
```
Comment (nafur) [inline src/smt/smt_engine.cpp:3206]:
```suggestion
void SmtEngine::setExpressionName(const Node& e, const std::string& name) {
```

File: src/smt/smt_engine.h [hunk 12]
```diff
@@ -858,15 +859,15 @@ class CVC4_PUBLIC SmtEngine
    * Return true if given expressoion has a name in the current context.
    * If it returns true, the name of expression 'e' is stored in 'name'.
    */
-  bool getExpressionName(Expr e, std::string& name) const;
+  bool getExpressionName(Node e, std::string& name) const;
```
Comment (nafur) [inline src/smt/smt_engine.h:862]:
```suggestion
  bool getExpressionName(const Node& e, std::string& name) const;
```

File: src/smt/smt_engine.h [hunk 13]
```diff
@@ -858,15 +859,15 @@ class CVC4_PUBLIC SmtEngine
...
 
   /**
    * Set name of given expression 'e' to 'name'.
    *
    * This information is user-context-dependent.
    * If 'e' already has a name, it is overwritten.
    */
-  void setExpressionName(Expr e, const std::string& name);
+  void setExpressionName(Node e, const std::string& name);
```
Comment (nafur) [inline src/smt/smt_engine.h:870]:
```suggestion
  void setExpressionName(const Node& e, const std::string& name);
```

File: src/smt/expr_names.h (hunk 1 above)
Comment (ajreynol) [reply src/smt/expr_names.h:33]: Yeah sorry I copied a template and forgot to clean.

File: src/smt/expr_names.cpp (hunk 7 above)
Comment (ajreynol) [reply src/smt/expr_names.cpp:30]: I don't like it, but I don't really care about this code.  I'll pick my battles ;)

**Pull request merged**